composer/loggers/neptune_logger.py

```diff
@@ -0,0 +1,316 @@
+
+        self._neptune_run = None
+        self._base_handler = None
+
+        self._metrics_dict: Dict[str, int] = {}  # used to prevent duplicate step logging
+
+        super(NeptuneLogger, self).__init__()
+
+    @property
+    def neptune_run(self):
+        """Gets the Neptune run object from a NeptuneLogger instance.
+
+        You can log additional metadata to the run by accessing a path inside the run and assigning metadata to it
+        with "=" or [Neptune logging methods](https://docs.neptune.ai/logging/methods/).
+
+        Example:
+                from composer import Trainer
+                from composer.loggers import NeptuneLogger
+                neptune_logger = NeptuneLogger()
+                trainer = Trainer(loggers=neptune_logger, ...)
+                trainer.fit()
+                neptune_logger.neptune_run["some_metric"] = 1
+                trainer.close()
+        """
+        from neptune import Run
+
+        if not self._neptune_run:
+            self._neptune_run = Run(
+                project=self._project,
+                api_token=self._api_token,
+                mode=self._mode,
+                **self._neptune_kwargs,
+            )
+        return self._neptune_run
+
+    @property
+    def base_handler(self):
+        """Gets a handler for the base logging namespace.
+
+        Use the handler to log extra metadata to the run and organize it under the base namespace (default: "training").
+        You can operate on it like a run object: Access a path inside the handler and assign metadata to it with "=" or
+        other [Neptune logging methods](https://docs.neptune.ai/logging/methods/).
+
+        Example:
+                from composer import Trainer
+                from composer.loggers import NeptuneLogger
+                neptune_logger = NeptuneLogger()
+                trainer = Trainer(loggers=neptune_logger, ...)
+                trainer.fit()
+                neptune_logger.base_handler["some_metric"] = 1
+                trainer.close()
+            Result: The value `1` is organized under "training/some_metric" inside the run.
+        """
+        return self.neptune_run[self._base_namespace]
+
+    def init(self, state: 'State', logger: 'Logger') -> None:
+        del logger  # unused
+
+        self.base_handler['rank'] = dist.get_global_rank()
+
+        if self._enabled:
+            self.neptune_run['sys/name'] = state.run_name
+            self.neptune_run[self.integration_version_key] = __version__
+
+    def _sanitize_metrics(self, metrics: Dict[str, float], step: Optional[int]) -> Dict[str, float]:
+        """Sanitize metrics to prevent duplicate step logging.
+
+        Args:
+            metrics (Dict[str, float]): Metrics to log.
+            step (Optional[int]): Step to log metrics at.
+
+        Returns:
+            Dict[str, float]: Sanitized metrics.
+        """
+        keys_to_delete: Set[str] = set()
+
+        for k in metrics:
```

Comment:
Can we extract the logic below to a separate private method or a function?

Reply:
Done